### PR TITLE
Accounts Refactor

### PR DESCRIPTION
**Pull request by alfiehanssen**
This is a major refactor of the networking library. Main components:

1. Accounts system is vastly simplified
2. VIMSession + VIMClient + VIMAuthenticator roles are far more straightforward
3. Dependencies drastically reduced

https://vimean.atlassian.net/browse/VIM-2446

### CODE REVIEW

File: VIMNetworking/Networking/VimeoClient/VIMRequestOperationManager.m [hunk 1]
```diff
@@ -58,28 +50,27 @@ @interface VIMRequestOperationManager ()
...
 {
+    NSParameterAssert(url);
+
     self = [super initWithBaseURL:url];
     if(self)
     {
-		_responseQueue = dispatch_queue_create("com.vimeo.VIMVimeoClient.responseQueue", DISPATCH_QUEUE_SERIAL);
+        // TODO: Do we need both queues or will one suffice? [AH]
+		
+        _responseQueue = dispatch_queue_create("com.vimeo.VIMClient.completionQueue", DISPATCH_QUEUE_SERIAL);
```
Comment (huebnerob) [inline VIMNetworking/Networking/VimeoClient/VIMRequestOperationManager.m]: com.vimeo.VIMClient.responseQueue?

File: VIMNetworking/Networking/VimeoClient/VIMSession.m [hunk 2]
```diff
@@ -25,478 +25,462 @@
...
-            return;
-        }
+    }];
+}
+
+- (id<VIMRequestToken>)authenticateWithCodeGrantResponseURL:(NSURL *)responseURL completionBlock:(VIMErrorCompletionBlock)completionBlock
+{
+    NSAssert([self.account isAuthenticatedWithUser] == NO, @"Attempt to authenticate as user when already authenticated as user");
+
+    typeof(self) weakSelf = self;
```
Comment (huebnerob) [inline VIMNetworking/Networking/VimeoClient/VIMSession.m]: __weak?

File: VIMNetworking/Networking/VimeoClient/VIMSession.m [hunk 3]
```diff
@@ -25,478 +25,462 @@
...
+        [weakSelf authenticationCompleteWithAccount:account error:error key:UserAccountKey completionBlock:completionBlock];
+
+    }];
+}
+
+- (id<VIMRequestToken>)loginWithEmail:(NSString *)email password:(NSString *)password completionBlock:(VIMErrorCompletionBlock)completionBlock
+{
+    NSAssert([self.account isAuthenticatedWithUser] == NO, @"Attempt to authenticate as user when already authenticated as user");
+
+    typeof(self) weakSelf = self;
```
Comment (huebnerob) [inline VIMNetworking/Networking/VimeoClient/VIMSession.m]: __weak here too?

File: VIMNetworking/Networking/VimeoClient/VIMSession.m [hunk 4]
```diff
@@ -25,478 +25,462 @@
...
+        [weakSelf authenticationCompleteWithAccount:account error:error key:UserAccountKey completionBlock:completionBlock];
+
+    }];
+}
+
+- (id<VIMRequestToken>)joinWithName:(NSString *)name email:(NSString *)email password:(NSString *)password completionBlock:(VIMErrorCompletionBlock)completionBlock
+{
+    NSAssert([self.account isAuthenticatedWithUser] == NO, @"Attempt to authenticate as user when already authenticated as user");
+
+    typeof(self) weakSelf = self;
```
Comment (huebnerob) [inline VIMNetworking/Networking/VimeoClient/VIMSession.m]: ^ again

File: VIMNetworking/Networking/VimeoClient/VIMSession.m [hunk 5]
```diff
@@ -25,478 +25,462 @@
...
     }];
 }
 
-#pragma mark - User Cache
+- (id<VIMRequestToken>)loginWithFacebookToken:(NSString *)facebookToken completionBlock:(VIMErrorCompletionBlock)completionBlock
+{
+    NSAssert([self.account isAuthenticatedWithUser] == NO, @"Attempt to authenticate as user when already authenticated as user");
 
-- (void)createUserCache
+    typeof(self) weakSelf = self;
```
Comment (huebnerob) [inline VIMNetworking/Networking/VimeoClient/VIMSession.m]: ^ __weak

File: VIMNetworking/Networking/VimeoClient/VIMSession.m [hunk 6]
```diff
@@ -25,478 +25,462 @@
...
-    if (self.authenticatedUser == nil)
-    {
-        _userCache = [VIMCache sharedCache];
-    }
-    else
-    {
-        NSString *userCacheName = [NSString stringWithFormat:@"user_%@", self.authenticatedUser.objectID];
+    NSAssert([self.account isAuthenticatedWithUser] == NO, @"Attempt to authenticate as user when already authenticated as user");
+
+    typeof(self) weakSelf = self;
```
Comment (huebnerob) [inline VIMNetworking/Networking/VimeoClient/VIMSession.m]: __weak

**Comment by huebnerob** [comment]
:+1:

File: VIMNetworking/Networking/VimeoClient/VIMRequestOperationManager.m (hunk 1 above)
Comment (alfiehanssen) [inline VIMNetworking/Networking/VimeoClient/VIMRequestOperationManager.m]: good catch on this and all of the below

**Pull request merged**